src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -306,28 +317,32 @@ private void OnSourceProviderSourcesChanged(object sender, EventArgs e)
             bool includePrerelease,
             Guid projectGuid,
             EnvDTE.DTE dte,
-            EnvDTE.Project dteProject)
+            EnvDTE.Project dteProject,
+            IProgressTracker progressTracker,
+            CancellationToken cancellationToken)
         {
             this.AssertIsForeground();
             Contract.ThrowIfFalse(IsEnabled);
 
-            // TODO: consider putting this under a TWD so it would actually be cancellable.
-            var cancellationToken = CancellationToken.None;
+            var description = string.Format(ServicesVSResources.Installing_0, packageName);
+            progressTracker.Description = description;
+            dte.StatusBar.Text = description;
 
             try
             {
-                return this.ThreadingContext.JoinableTaskFactory.Run(() =>
-                {
-                    return this.PerformNuGetProjectServiceWorkAsync(async (nugetService, cancellationToken) =>
+                return this.ThreadingContext.JoinableTaskFactory.Run(
```

Comment:
Is now a good time to consider whether blocking the caller during this async work is really necessary?
You remove a SwitchToMainThreadAsync and add a switch to the threadpool, but if this code runs while a JTF.Run method is blocking our caller (who we verify above *is* on the main thread), then it's all moot because you'll block the main thread during the whole thing anyway.

Reply:
the main issue here is that this is called from the undo/redo stack.  and it hurts my brain to try to consider what it means to do undo/redo asynchronously.